patches/server/1040-Add-raw-byte-block-state-serialization.patch

```diff
@@ -0,0 +1,77 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: TrollyLoki <trollyloki@gmail.com>
+Date: Thu, 12 Oct 2023 11:05:19 -0400
+Subject: [PATCH] Add raw byte block state serialization
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
+index e3b07d623cd64de9645f2372f1e08757edc1a9ed..c55969167cdb30a926158756c1ac21e489cde7e9 100644
+--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
++++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
+@@ -231,13 +231,17 @@ public final class CraftBlockStates {
+         return CraftBlockStates.getBlockState(block, true);
+     }
+     public static BlockState getBlockState(Block block, boolean useSnapshot) {
++        return CraftBlockStates.getBlockState(block, useSnapshot, null, null);
++    }
++    // customTileEntity is ignored if customBlockData is null
++    public static BlockState getBlockState(Block block, boolean useSnapshot, net.minecraft.world.level.block.state.BlockState customBlockData, BlockEntity customTileEntity) {
+         // Paper end
+         Preconditions.checkNotNull(block, "block is null");
+         CraftBlock craftBlock = (CraftBlock) block;
+         CraftWorld world = (CraftWorld) block.getWorld();
+         BlockPos blockPosition = craftBlock.getPosition();
+-        net.minecraft.world.level.block.state.BlockState blockData = craftBlock.getNMS();
+-        BlockEntity tileEntity = craftBlock.getHandle().getBlockEntity(blockPosition);
++        net.minecraft.world.level.block.state.BlockState blockData = customBlockData != null ? customBlockData : craftBlock.getNMS(); // Paper - use customBlockData if provided
++        BlockEntity tileEntity = customBlockData != null ? customTileEntity : craftBlock.getHandle().getBlockEntity(blockPosition); // Paper - use customTileEntity if customBlockData provided
```

Comment:
Should we be checking both customBlockData and customTileEntity to not be null here?
Right now it is a pretty implicit contract to pass either both null or both not null.

Reply:
If a deserialized block does not have an associated tile entity, the passed customTileEntity will be null but customBlockData will not be. I am essentially using the non-null status of customBlockData as a flag to indicate that it should use the customTileEntity be it null or non-null instead of getting the tile entity from the world.
Would it be better to have a separate boolean argument specifying which block data and tile entity should be used?

Reply:
Hmmm, no I think this is fine then. Would probably update the comment above the method talking about this to be a bit more specific then.

I'll go through a deeper review this weekend (tomorrow hopefully), but yea, should be fine then.